From d5f5974c810a9a540f507a5ede5fafb56132e94c Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 25 Jul 2024 07:19:54 +0300
Subject: [PATCH] [FSW-1175] fs1: recovery: boot recovery img

When we enter recovery if both manifests are invalid
then it should execute rom recovery. Otherwise it
should boot recovery image from primary/secondary
manifest.

Change-Id: I40f2c6fb077fbbcd28d89ea736ce5ddb1902d62f
---

diff --git a/common/recovery.c b/common/recovery.c
index f7e23f4..8974f9e 100644
--- a/common/recovery.c
+++ b/common/recovery.c
@@ -6,6 +6,9 @@
 
 #include <boot_rom.h>
 #include <recovery.h>
+#include <boot_rom.h>
+#include <boot_flow.h>
+#include <common.h>
 #include <fs1_reg_map.h>
 #include <fs1_blocks_linux_driver.h>
 #include <global_conf_masks.h>
@@ -27,6 +30,11 @@
 	boot_rom_status_update(PROGRESS_BOOT_ROM_RECOVERY_END);
 }
 
+static int boot_rom_execute_recovery_img(__unused enum boot_rom_manifests recovery_manifest)
+{
+	return BOOT_ROM_SUCCESS;
+}
+
 void execute_recovery(void)
 {
 	boot_rom_status_update(PROGRESS_BOOT_ROM_IN_RECOVERY);
@@ -36,5 +44,22 @@
 	if (boot_rom_in_multi_die_mode())
 		boot_rom_notify_slave();
 
+	if (!boot_rom_boot_flow_is_manifest_valid(BOOT_ROM_PRIMARY) &&
+	    !boot_rom_boot_flow_is_manifest_valid(BOOT_ROM_SECONDARY))
+		goto ROM_RECOVERY;
+
+	if (boot_rom_boot_flow_is_manifest_valid(BOOT_ROM_PRIMARY)) {
+		if (boot_rom_execute_recovery_img(BOOT_ROM_PRIMARY) == BOOT_ROM_SUCCESS)
+			goto HALT;
+	}
+
+	if (boot_rom_boot_flow_is_manifest_valid(BOOT_ROM_SECONDARY)) {
+		if (boot_rom_execute_recovery_img(BOOT_ROM_SECONDARY) == BOOT_ROM_SUCCESS)
+			goto HALT;
+	}
+
+ROM_RECOVERY:
 	execute_rom_recovery();
+HALT:
+	boot_rom_halt();
 }
